plugins/statistics.koplugin/main.lua

```diff
@@ -533,47 +684,59 @@ function ReaderStatistics:getIdBookDB()
-    self.pages_stats[TimeVal:now().sec] = self.curr_page
-    if id_book == nil or util.tableSize(self.pages_stats) < 2 then
+function ReaderStatistics:insertDB(id_book, updated_pagecount)
+    if id_book == nil or self.mem_read_pages < MIN_PAGES_READ_TO_FLUSH then
         return
     end
-    local diff_time
+    local now_ts = TimeVal:now().sec
     local conn = SQ3.open(db_location)
-    local sorted_performance = {}
-    for time, pages in pairs(self.pages_stats) do
-        table.insert(sorted_performance, time)
-    end
-    table.sort(sorted_performance)
-    conn:exec('BEGIN')
-    local stmt = conn:prepare("INSERT OR IGNORE INTO page_stat VALUES(?, ?, ?, ?)")
-    for i=1, #sorted_performance - 1 do
-        diff_time = sorted_performance[i+1] - sorted_performance[i]
-        if diff_time >= self.page_min_read_sec then
-            stmt:reset():bind(id_book,
-                self.pages_stats[sorted_performance[i]],
-                sorted_performance[i],
-                math.min(diff_time, self.page_max_read_sec)):step()
+    conn:exec('BEGIN;')
+    local stmt = conn:prepare("INSERT OR IGNORE INTO page_stat_data VALUES(?, ?, ?, ?, ?);")
+    for page, data_list in pairs(self.page_stat) do
+        for _, data_tuple in ipairs(data_list) do
+            -- See self.page_stat declaration above about the tuple's layout
+            local ts = data_tuple[1]
+            local duration = data_tuple[2]
+            -- Skip placeholder durations
+            if duration > 0 then
+                -- NOTE: The fact that we update self.data.pages *after* this call on layout changes
+                --       should ensure that it matches the layout in which said data was collected.
+                --       Said data is used to re-scale page numbers, regardless of the document layout,
+                --       at query time, via a fancy SQL view.
+                --       This allows the progress tracking to be accurate even in the face of wild
+                --       document layout changes (e.g., after font size changes).
+                stmt:reset():bind(id_book, page, ts, duration, self.data.pages):step()
+            end
         end
     end
-    conn:exec('COMMIT')
+    conn:exec('COMMIT;')
+    -- NOTE: See the tail end of the discussions in #6761 for more context on the choice of this heuristic.
+    --       Basically, we're counting distinct pages,
+    --       while making sure the sum of durations per distinct page is clamped to self.page_max_read_sec
+    --       This is expressly tailored to a fairer computation of self.avg_time ;).
     local sql_stmt = [[
-        SELECT count(DISTINCT page),
-               sum(period)
-        FROM   page_stat
-        WHERE  id_book = '%s'
+        SELECT count(*),
+               sum(durations)
+        FROM (
+            SELECT min(sum(duration), %d) AS durations
+            FROM page_stat_data
+            WHERE id_book = %d
+            GROUP BY page
```

Comment:
As I'm ok with the count(*), we have definitely forget here about the `self.page_min_read_sec`. (the funny thing is that page_min_read_sec is not working correctly even on the last official release build).

I think the we should check this in the HAVING statement (per every entry, not on the grouped by). And also the change with min(sum()) to sum(min()) with page_max_read_sec - see what's going on if we read page twice (total_pages = 2, read_time = once? - this is wrong).

I propose:

SELECT sum(min(duration, %d)) AS durations --- sum(min())
            FROM page_stat_data
            WHERE id_book = %d
            GROUP BY page
            HAVING duration > %d -- page_min_read_sec

Sorry for that at it was mine quick sql example :)

Reply:
The data collection handler ensures that *nothing* will ever be collected below `self.page_min_read_sec`.

... at least as far as `self.page_min_read_sec` is concerned *at the time of collection* ;).

Adding a constraint here would essentially only be useful if `self.page_min_read_sec` was modified in the middle of reading *this* book.

Which is fine by me, but may not actually be all that useful in practice ^^.

Reply:
Ok. And what about the second part: "And also the change with min(sum()) to sum(min()) with page_max_read_sec - see what's going on if we read page twice (total_pages = 2, read_time = once? - this is wrong)." ?

Reply:
Ok. And what about the second part: "And also the change with min(sum()) to sum(min()) with page_max_read_sec - see what's going on if we read page twice (total_pages = 2, read_time = once? - this is wrong)." ?